charger/peblar.go

```diff
@@ -0,0 +1,304 @@
+	return wb.MaxCurrentMillis(float64(current))
+}
+
+var _ api.ChargerEx = (*Peblar)(nil)
+
+// MaxCurrent implements the api.ChargerEx interface
+func (wb *Peblar) MaxCurrentMillis(current float64) error {
+	if current < 6 {
+		return fmt.Errorf("invalid current %.1f", current)
+	}
+
+	err := wb.setCurrent(wb.curr)
+	if err == nil {
+		wb.curr = uint32(current * 1e3)
+	}
+
+	return err
+}
+
+var _ api.Meter = (*Peblar)(nil)
+
+// CurrentPower implements the api.Meter interface
+func (wb *Peblar) CurrentPower() (float64, error) {
+	b, err := wb.conn.ReadInputRegisters(peblarPowerTotalAddress, 2)
+	if err != nil {
+		return 0, err
+	}
+
+	return float64(binary.BigEndian.Uint32(b)), err
+}
+
+var _ api.ChargeRater = (*Peblar)(nil)
+
+// ChargedEnergy implements the api.MeterEnergy interface
+func (wb *Peblar) ChargedEnergy() (float64, error) {
+	b, err := wb.conn.ReadInputRegisters(peblarSessionEnergyAddress, 4)
+	if err != nil {
+		return 0, err
+	}
+
+	return float64(encoding.Int64(b)) / 1e3, err
+}
+
+// TotalEnergy implements the api.MeterEnergy interface
+func (wb *Peblar) TotalEnergy() (float64, error) {
+	b, err := wb.conn.ReadInputRegisters(peblarEnergyTotalAddress, 4)
+	if err != nil {
+		return 0, err
+	}
+
+	return float64(encoding.Int64(b)) / 1e3, nil
+}
+
+// getPhaseValues returns 1..3 sequential register values
+func (wb *Peblar) getPhaseValues(reg uint16, divider float64) (float64, float64, float64, error) {
+	b, err := wb.conn.ReadInputRegisters(reg, wb.phases*2)
+	if err != nil {
+		return 0, 0, 0, err
+	}
+
+	var res [3]float64
+	for i := 0; i < int(wb.phases); i++ {
+		res[i] = float64(binary.BigEndian.Uint32(b[4*i:])) / divider
+	}
+
+	return res[0], res[1], res[2], nil
+}
+
+var _ api.PhaseCurrents = (*Peblar)(nil)
+
+// Currents implements the api.PhaseCurrents interface
+func (wb *Peblar) Currents() (float64, float64, float64, error) {
+	return wb.getPhaseValues(peblarCurrentPhase1Address, 1e3)
+}
+
+var _ api.PhaseVoltages = (*Peblar)(nil)
+
+// Voltages implements the api.PhaseVoltages interface
+func (wb *Peblar) Voltages() (float64, float64, float64, error) {
+	return wb.getPhaseValues(peblarVoltagePhase1Address, 1)
+}
+
+// phases1p3p implements the api.PhaseSwitcher interface via the decorator
+func (wb *Peblar) phases1p3p(phases int) error {
+	var b uint16
+	if phases == 1 {
+		b = 1
+	}
+
+	_, err := wb.conn.WriteSingleRegister(peblarForce1PhaseAddress, b)
+	return err
+}
+
+var _ api.PhaseGetter = (*Peblar)(nil)
+
+// GetPhases implements the api.PhaseGetter interface
+func (wb *Peblar) GetPhases() (int, error) {
+	c, err := wb.conn.ReadHoldingRegisters(peblarForce1PhaseAddress, 1)
+	if err != nil {
+		return 0, err
+	}
+
+	var force1p uint16 = binary.BigEndian.Uint16(c)
+	wb.log.DEBUG.Println("forced to 1 phase:", force1p)
+	var phases uint16 = wb.phases
+	wb.log.DEBUG.Println("connected phases:", phases)
+
+	if wb.phases > 1 && force1p == 1 {
+		phases = 1
+	}
+
+	return int(phases), nil
```

Comment:
This should return the switched phases- 1 or 3. If config is fixed to 1p connected switching will never be invoked. Why do we need this complexity here?

Reply:
Clear..
2 relays, 3phase connected:  Both GetPhases & phase1p3p supported.
1 or 2 relays, 1phase connected: No GetPhases, no phase1p3p.
1 relay, 3phase connected: GetPhases possible but phase1p3p not.

Reply:
> But what about the option where we switch all phases with 1 relay? Clearly no phase1p3p, but also no GetPhases?

Not quite sure what you mean. If box is connected 1p that's part of loadpoint config that we rely on. GetPhases will only every be called if Phases1p3p is actually available.